metaflow/plugins/argo/event_decorators.py

```diff
@@ -0,0 +1,247 @@
+                )
+            trigger_on = {
+                "flow": flow_name,
+                "status": None,
+                "event": event_name,
+                "mappings": None,
+            }
+            if flow_name is not None:
+                status = self.attributes.get("status").lower()
+                if status not in valid_statuses():
+                    raise MetaflowException(
+                        msg=(
+                            "@%s requires status to be one of: %s."
+                            % (self.name, ", ".join(valid_statuses()))
+                        )
+                    )
+                trigger_on["status"] = status
+            elif event_name is not None:
+                sanitized = sanitize_user_event(event_name)
+                if sanitized == "":
+                    raise BadEventNameException(BAD_EVENT_ERR_MSG)
+                trigger_on["event"] = sanitized
+            mappings = self.attributes.get("mappings")
+            trigger_on["mappings"] = validate_mappings(mappings)
+            current._update_env({"trigger_on": trigger_on})
+
+        else:
+            # Defer raising an error in case user has specified --ignore-triggers
+            current._update_env(
+                {
+                    "event_decorator_error": {
+                        "message": ERR_MSG,
+                        "headline": ("@%s requires eventing support" % self.name),
+                    }
+                }
+            )
+
+
+class EmitEventDecorator(StepDecorator):
+
+    name = "emit_event"
+    defaults = {"event": None, "task_status": "succeeded", "data": {}}
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        if self.attributes.get("task_status") not in valid_statuses():
+            raise MetaflowException(
+                msg=(
+                    "@%s requires task_status to be one of: %s."
+                    % (self.name, ", ".join(valid_statuses()))
+                )
+            )
+        event_name = self.attributes.get("event")
+        if event_name in [None, ""]:
+            raise BadEventNameException("@%s needs an event" % self.name)
+        event_name = sanitize_user_event(event_name)
+        if event_name == "":
+            raise BadEventNameException(BAD_EVENT_ERR_MSG)
+        self.attributes["event"] = event_name
+        data = self.attributes.get("data")
+        if data is not None:
+            validated = validate_data(self.name, data)
+            self.attributes["data"] = validated
+
+    def formatted_status(self):
+        status = self.attributes.get("task_status")
+        if status is None:
+            return None
+        return status.capitalize()
+
+    @property
+    def event(self):
+        return self.attributes["event"]
+
+    @property
+    def data(self):
+        return self.attributes.get("data")
+
+
+class AnnotateLifecycleDecorator(FlowDecorator):
```

Comment:
What is the intended use case for this?

Reply:
it is quite likely that we may need some customization for Argo Workflows - let's chat with the [Deliveroo folks](https://outerbounds-community.slack.com/archives/C02116BBNTU/p1663344880751489?thread_ts=1663253873.389359&cid=C02116BBNTU) independent of this effort.